### ISSUE TITLE
No sound - dolby with headset connected - when switch from DVB playback to OTT (ExoPlayer)

### ISSUE THREAD

**Issue by ZygmuntMalek** [androidx/media#440]
### Media3 Version

ExoPlayer 2.18.6

### Devices that reproduce the issue

Amino hybrid box 

### Reproducible in the demo app?

Not tested

### Reproduction steps

Precondition
There is dolby audio stream (codec: ac-3). Device has audio decoder, There is connected headset by Bluetooth 

1 Step) Play OTT stream - select dolby audio track, track is down-mixed, playback could be heard.
2 Step) Play DVB stream 
3 Step) Stop DVB stream
4 Step) Play OTT stream - select dolby audio track, there is error in logs like 

> 06-01 10:09:22.301  7551  9033 I BTAudioProviderStub: streamStarted - SessionType=A2DP_SOFTWARE_ENCODING_DATAPATH, status=FAILURE
06-01 10:09:22.301  7551  9033 I BTAudioProviderSession: ReportControlStatus - status=FAILURE for SessionType=A2DP_SOFTWARE_ENCODING_DATAPATH, bluetooth_audio=0x0100 started
06-01 10:09:22.301  7551  9033 D A2DPHW  : BluetoothAudioPortOut::Start: state=1, ret=1
06-01 10:09:22.303  7551  9033 D a2dp_hal: a2dp_out_resume: state=1

### Expected result

if stream can not be down-mixed there should be reject when selecting dolby track, or supported flag should be to false

Also this looks like regression between versions as issue is not reproducible at version 2.12.3
It's reproducible at 2.17.1 ( which we currently use, as also as we tested it's still reproducible at version 2.18.6 )

### Actual result

There is no sound, there is no error notification from ExoPlayer

Attached bug report from version 2.18.6

### Media

no specific media

### Bug Report

- [X] You will email the zip file produced by `adb bugreport` to dev.exoplayer@gmail.com after filing this issue.

**Comment by oceanjules** [comment]
Hi, thank you for reporting the issue. In order for us to reproduce it, would you mind sharing the media you were using for it?

Also, when you say "DVB playback", do you mean an RF signal carrying DVB? And is that arriving on the device via TV Input Framework?

**Comment by ZygmuntMalek** [comment]
Yes, signal DVB is from RF input. I do not have knowledge about solution over DVB as it's delivered by device manufacturer as third party library (we have support of manufacturer but communication pipe is long ). 
Used media do not looks like important in this case.
From my perspective it looks like as there in an issue with using device decoder to down-mix Dolby track ( Manufacturer reported that not down-mixed stream were tried to pushed by Bluetooth )

**Comment by christosts** [comment]
It will be hard to investigate without a way to reproduce locally. I understand this is specific to the device and perhaps to the content or the way the app is using ExoPlayer.

> From my perspective it looks like as there in an issue with using device decoder to down-mix Dolby track ( Manufacturer reported that not down-mixed stream were tried to pushed by Bluetooth )

ExoPlayer does not know whether the Dolby decoder will down-mix surround audio. On Android 12L+, an the app can ask the decoder to not downmix. ExoPlayer does ask the decoder to not downmix as of 2.18.0. However, you observe issues with prior ExoPlayer versions and I don't think the Amino hybrid box is on Android 12L or later. In addition, even if the decoder outputs multi-channel PCM, the platform (=> AudioTrack) should down-mix and transmit stereo to the BT headset. 

Without a way to repro, I can only ask for more information to narrow-down the conditions the issue occurs:

1. Is the error happening only when you switch between OTT -> DVB -> OTT?
2. What is the audio of the DVB stream? AAC stereo?
3. Can the error be reproduced in the ExoPlayer demo app, e.g. switching between AC3 -> [the same audio as the DVB stream] -> AC3 streams?
4. If yes, can you please repro on the ExoPlayer demo app and send us a bug-report after the error occurs? It is critical to use the demo app because we know how the app uses ExoPlayer and it produces additional logs.
5. If you cannot use the demo app, can you please attach an `EventLogger` on the ExoPlayer instances (with `Player.addAnalyticsListener`) on your app and get another bug-report? Note that the bug-report we received has no logs coming out from ExoPlayer. If you've disabled ExoPlayer's logging, please re-enable them.

Thanks

**Comment by ZygmuntMalek** [comment]
I've send another bug report with EventLogger enabled, I've also attached logs gathered by adb logcat.
Ad1 ) Maybe my description were misleading 
When OTT is started directly bug is not reproduced, to trigger it before OTT there is need to do DVB playback

Ad2) DVB audio is MPEG 1 - Layer II

Ad3,4 ) Establish such environment would be a bit difficult for us & imho this would not reproduce issue, we will try it if other analysis will fail.

About device specification I think it could be read from bug report, If you will have some issue with that, please let me know I'll get it for you

**Comment by christosts** [comment]
Thanks for sending the BR with the additional logs. May I ask, at what point in your setup do you connect to BT headset? Is it always connected or do you connect it *after* already having played the OTT the first time?

TLDR: I believe we have a fix for this issue but it's not yet released. It will be in the upcoming media3 1.1.0 (or ExoPlayer 2.19.0). To verify the fix addressed your issue, can you build you app using ExoPlayer from the dev-v2 branch (or media3 from main branch)? You need to make your app depends on the library [locally](https://github.com/androidx/media/blob/release/README.md#locally).

More details: There is a subtle detail in the player logs. While the device can play ac-3, the platform is not exposing an ac-3 decoder, rather the device reports it can play ac-3 directly via passthrough. I'm not sure if the device has an ac-3 hardware decoder that is only exposed via passthrough, or the device is plugged over HDMI to a TV (or AV receiver) that supports ac-3, but that's a detail.  We have a fix on the player to detect that when the device is connected to a BT headset, the player should not play surround formats that are supported via passthrough and either play stereo or use a track that is decoded locally to PCM (the platform will downminx to stereo before transmitting to the BT headset). The player fix also detects when a BT headset is connected on the device mid-playback and the player should automatically switch from ac-3 to a compatible audio track.

I noticed however the following: on your device, when the player attempts to play ac-3 over BT, there's a native crash. We expect to get an error at the Java-layer and to act upon it. This is going to impact the case where the player is already playing an ac-3 and the user connects the BT headset midstream. The platform should not crash, but we expect to receive an error and  switch to stereo. This is not something ExoPlayer can control.

**Comment by christosts** [comment]
For reference, I think this issue is likely a duplicate of either https://github.com/google/ExoPlayer/issues/9588 or https://github.com/google/ExoPlayer/issues/10227.

**Comment by ZygmuntMalek** [comment]
I can confirm that using build from dev branch resolves our client issues

**Comment by christosts** [comment]
Glad to hear. media3-1.1.0 is already in beta, and we expect to release media3-1.0.0 stable sometime in July. ExoPlayer 2.19.0 will be released with media3-1.0.0 stable.

I'm closing this issue as the underlying issue is fixed.
